### PR TITLE
topology: sof-adl-cs35l41: support cs35l41 speaker amplifier

### PR DESCRIPTION
**Pull request by brentlu**
Add a topology sof-adl-cs35l41 which supports four Cirrus Logic
CS35L41/CS35L53 speaker amplifiers without headphone codec on ADL
platform.

Signed-off-by: Brent Lu <brent.lu@intel.com>

### CODE REVIEW

**Comment by brentlu** [comment]
Linux PR: https://github.com/thesofproject/linux/pull/3436

**Review by lgirdwood — COMMENTED** [review]
@brentlu I assume you will be validating the BT part ?
Adding @kv2019i and @plbossart for TDM/BT inputs.

**Comment by brentlu** [comment]
This topology is not so ready yet. We plan to add BT offload support and change the DAI to TDM mode for feedback from 4 amps. However, I'm still waiting for a stable SW/HW platform to test it...

**Comment by gkbldcig** [comment]
Can one of the admins verify this patch?

**Comment by brentlu** [comment]
> Can one of the admins verify this patch?

I will work with ODM to verify the machine driver and topology on their device then update the result here.

**Comment by plbossart** [comment]
> Can one of the admins verify this patch?

can one of the QB admins verify why we need to verify the patch of an SOF project member?

**Comment by lgirdwood** [comment]
@brentlu any update ?

**Comment by brentlu** [comment]
Tested on Chromebook for speaker output.

**Comment by brentlu** [comment]
Corresponding Linux PR:

**Comment by lgirdwood** [comment]
@brentlu lets merge this after the kernel PR is aligned. Ping me when kernel PR is merged.

File: tools/topology/topology1/sof-tgl-max98357a-rt5682.m4 [hunk 1]
```diff
@@ -426,24 +448,31 @@ ifelse(
...
 		SSP_TDM(2, 25, 3, 3),
-		SSP_CONFIG_DATA(SSP, 0, 24, 0, 0, 0, SSP_CC_BCLK_ES)))
+		SSP_CONFIG_DATA(SSP, 0, 24, 0, 0, 0, SSP_CC_BCLK_ES)))')
 
 # 4 HDMI/DP outputs (ID: 3,4,5,6)
-DAI_CONFIG(HDA, 0, 3, iDisp1,
+DAI_CONFIG(HDA, 0, HDMI_BE_ID_BASE, iDisp1,
 	HDA_CONFIG(HDA_CONFIG_DATA(HDA, 0, 48000, 2)))
-DAI_CONFIG(HDA, 1, 4, iDisp2,
+DAI_CONFIG(HDA, 1, eval(HDMI_BE_ID_BASE + 1), iDisp2,
```
Comment (cujomalainey) [inline tools/topology/topology1/sof-tgl-max98357a-rt5682.m4:471]: i wonder if we should just make this its own M4 we can include with a parameter to set the base, would save some code space for all boards.

**Comment by brentlu** [comment]
Add a compile option 'SSP_AMP_LINK_ORDER' to use the BE DAI Link order in sof_ssp_amp.c

File: tools/topology/topology1/sof-tgl-max98357a-rt5682.m4 (hunk 1 above)
Comment (lgirdwood) [reply tools/topology/topology1/sof-tgl-max98357a-rt5682.m4:479]: yeah, having a separate M4 file that could be included here with the ID definitions for each board would make the maintenance easier.

File: tools/topology/topology1/platform/intel/intel-generic-hdmi.m4 [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+# Schedule 48 frames per 1000us deadline with priority 0 on core 0
+PIPELINE_PCM_ADD(sof/pipe-volume-playback.m4,
+	HDMI_PCM_ID_BASE, HDMI_PIPELINE_ID_BASE, 2, s32le,
+	1000, 0, 0,
+	48000, 48000, 48000)
+
+`# Low Latency playback pipeline 'eval(HDMI_PIPELINE_ID_BASE + 1)` on PCM 'eval(HDMI_PCM_ID_BASE + 1)` using max 2 channels of s32le.'
+# Schedule 48 frames per 1000us deadline with priority 0 on core 0
+PIPELINE_PCM_ADD(sof/pipe-volume-playback.m4,
+	eval(HDMI_PCM_ID_BASE + 1), eval(HDMI_PIPELINE_ID_BASE + 1), 2, s32le,
```
Comment (lgirdwood) [inline tools/topology/topology1/platform/intel/intel-generic-hdmi.m4:44]: sorry, I'meant a separate M4 to define the pipeline IDs based on the base ID. Is this how it is done in other files ?

File: tools/topology/topology1/platform/intel/intel-boards.m4 [hunk 3]
```diff
@@ -0,0 +1,34 @@
+#
+# Machine driver differentiation for dai link ID
+#
+
+# variable that need to be defined in upper m4
+ifdef(`PLATFORM',`',`fatal_error(note: Need to define platform idendifier for intel-boards)')
+ifdef(`BOARD',`',`fatal_error(note: Need to define board idendifier for intel-boards)')
```
Comment (plbossart) [inline tools/topology/topology1/platform/intel/intel-boards.m4:7]: typo: identifier

File: tools/topology/topology1/platform/intel/intel-boards.m4 [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+ifelse(
+	PLATFORM, `cml', `define(`NUM_HDMI_BE_LINK', 3)', 
+	PLATFORM, `jsl', `define(`NUM_HDMI_BE_LINK', 3)', 
+	PLATFORM, `tgl', `define(`NUM_HDMI_BE_LINK', 4)', 
+	PLATFORM, `adl', `define(`NUM_HDMI_BE_LINK', 4)',
+	`fatal_error(note: Unknown platform to intel-boards)')
+
+# define the ID base for backend DAI Links
+ifelse(
+	BOARD, `sof_rt5682', `
```
Comment (plbossart) [inline tools/topology/topology1/platform/intel/intel-boards.m4:20]: It's not really a BOARD, it's a machine driver that you are referring to.

LINUX_MACHINE_DRIVER would be non-ambiguous.

File: tools/topology/topology1/platform/intel/intel-boards.m4 (hunk 3 above)
Comment (plbossart) [reply tools/topology/topology1/platform/intel/intel-boards.m4:7]: also typo in commit title: s/machind/machine/

File: tools/topology/topology1/CMakeLists.txt [hunk 5]
```diff
@@ -132,30 +132,30 @@ set(TPLGS
...
+	"sof-tgl-max98357a-rt5682\;sof-adl-max98390-rt5682\;-DCODEC=MAX98390\;-DFMT=s32le\;-DPLATFORM=adl\;-DBOARD=sof_rt5682\;-DAMP_SSP=1\;-DBT_OFFLOAD"
+	"sof-tgl-max98357a-rt5682\;sof-adl-max98390-rt5682-google-aec\;-DCODEC=MAX98390\;-DFMT=s32le\;-DPLATFORM=adl\;-DBOARD=sof_rt5682\;-DAMP_SSP=1\;-DBT_OFFLOAD\;-DGOOGLE_RTC_AUDIO_PROCESSING"
+	"sof-tgl-max98357a-rt5682\;sof-adl-max98390-rt5682-rtnr\;-DCODEC=MAX98390\;-DFMT=s32le\;-DPLATFORM=adl\;-DBOARD=sof_rt5682\;-DAMP_SSP=1\;-DBT_OFFLOAD\;-DCHANNELS=2\;-DRTNR\;-DNOHOTWORD\;-DDMICPROC=rtnr\;-DNO16KDMIC\;-DDMIC_48k_CORE_ID=1"
+	"sof-tgl-max98357a-rt5682\;sof-adl-max98390-ssp2-rt5682-ssp0\;-DCODEC=MAX98390\;-DFMT=s32le\;-DPLATFORM=adl\;-DBOARD=sof_rt5682\;-DAMP_SSP=2"
+	"sof-tgl-max98357a-rt5682\;sof-adl-rt5682\;-DNO_AMP\;-DPLATFORM=adl\;-DBOARD=sof_rt5682"
+	"sof-tgl-max98357a-rt5682\;sof-adl-rt1019-rt5682\;-DCODEC=RT1019\;-DFMT=s16le\;-DPLATFORM=adl\;-DBOARD=sof_rt5682\;-DAMP_SSP=1"
+	"sof-tgl-max98357a-rt5682\;sof-adl-cs35l41\;-DCODEC=CS35L41\;-DFMT=s24le\;-DPLATFORM=adl\;-DBOARD=sof_ssp_amp\;-DAMP_SSP=1\;-DBT_OFFLOAD\;-DNO_HEADPHONE"
```
Comment (plbossart) [inline tools/topology/topology1/CMakeLists.txt:158]: I think you need to make sure this ssp_amp order works even without dmics.

It's fine to assume, as you did that the dmics are present, but to future-proof we should add a -DNO_DMICS option to make sure the order works in all cases.

**Review by plbossart — COMMENTED** [review]
Looks mostly good @brentlu, you have a couple of typos and I'd like the ssp_amp support to be improved when there are no mics, but otherwise it's going in the right direction. Nice work.

File: tools/topology/topology1/platform/intel/intel-boards.m4 (hunk 4 above)
Comment (brentlu) [reply tools/topology/topology1/platform/intel/intel-boards.m4]: fixed

File: tools/topology/topology1/platform/intel/intel-boards.m4 (hunk 3 above)
Comment (brentlu) [reply tools/topology/topology1/platform/intel/intel-boards.m4]: fixed

File: tools/topology/topology1/platform/intel/intel-generic-hdmi.m4 (hunk 2 above)
Comment (brentlu) [reply tools/topology/topology1/platform/intel/intel-generic-hdmi.m4]: I added a intel-boards.m4 to define the dai link ids which are supposed to match the same id in machine driver

**Comment by lgirdwood** [comment]
@plbossart good for you now ?

**Review by plbossart — APPROVED** [review]
Thanks @brentlu for addressing my comments, this looks good now.

@lgirdwood this is ready to merge.

**Review by cujomalainey — APPROVED** [review]

**Pull request merged**